Intersect (Core)/GameObjects/UserVariableBase.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Linq;
+using Intersect.Enums;
+using Intersect.Models;
+
+using Newtonsoft.Json;
+
+namespace Intersect.GameObjects
+{
+    public partial class UserVariableBase : DatabaseObject<UserVariableBase>, IFolderable
+    {
+
+        [JsonConstructor]
+        public UserVariableBase(Guid id) : base(id)
+        {
+            Name = "New Account Variable";
+        }
+
+        public UserVariableBase()
+        {
+            Name = "New Account Variable";
+        }
+
+        //Identifier used for event chat variables to display the value of this variable/switch.
+        //See https://www.ascensiongamedev.com/topic/749-event-text-variables/ for usage info.
+        public string TextId { get; set; }
+
+        public VariableDataTypes Type { get; set; } = VariableDataTypes.Boolean;
+
+        /// <inheritdoc />
+        public string Folder { get; set; } = "";
+
+        /// <summary>
+        /// Retrieve an array of variable names of the supplied data type.
+        /// </summary>
+        /// <param name="dataType">The data type to retrieve names of.</param>
+        /// <returns>Returns an array of names.</returns>
+        public static string[] GetNamesByType(VariableDataTypes dataType)
+        {
+            return Lookup.KeyList.OrderBy(pairs => Lookup[pairs]?.TimeCreated).Where(pairs => ((UserVariableBase)Lookup[pairs]).Type == dataType).Select(pairs => ((UserVariableBase)Lookup[pairs]).Name).ToArray();
+        }
+
+        /// <summary>
+        /// Retrieve the list index of an Id within a specific data type list.
+        /// </summary>
+        /// <param name="id">The Id to look up.</param>
+        /// <param name="dataType">The data type to search up.</param>
+        /// <returns>Returns the list Index of the provided Id.</returns>
+        public static int ListIndex(Guid id, VariableDataTypes dataType)
+        {
+            return Lookup.KeyList.OrderBy(pairs => Lookup[pairs]?.TimeCreated).Where(pairs => ((UserVariableBase)Lookup[pairs]).Type == dataType).Select(pairs => ((UserVariableBase)Lookup[pairs]).Id).ToList().IndexOf(id);
+        }
+
+        /// <summary>
+        /// Retrieve the Id associated with a list index of a specific data type.
+        /// </summary>
+        /// <param name="listIndex">The list index to retrieve.</param>
+        /// <param name="dataType">The data type to search up.</param>
+        /// <returns>Returns the Id of the provided index.</returns>
+        public static Guid IdFromList(int listIndex, VariableDataTypes dataType)
+        {
+            if (listIndex < 0 || listIndex > GetNamesByType(dataType).Length)
+            {
+                return Guid.Empty;
+            }
+
+            return Lookup.KeyList.OrderBy(pairs => Lookup[pairs]?.TimeCreated).Where(pairs => ((UserVariableBase)Lookup[pairs]).Type == dataType).Select(pairs => ((UserVariableBase)Lookup[pairs]).Id).ToArray()[listIndex];
+        }
+    }
```

Comment:
Ok you based this on the original Guild Variables implementation which is horrifyingly bad for performance.

Model your code off the following rewrite of the Guild Variable logic, replacing Type with DataType like we discussed on Discord:

```csharp
        /// <summary>
        /// Retrieve an array of variable names of the supplied data type.
        /// </summary>
        /// <param name="dataType">The data type to retrieve names of.</param>
        /// <returns>Returns an array of names.</returns>
        public static string[] GetNamesByType(VariableDataTypes dataType) =>
            Lookup
                .Where(pair => pair.Value is GuildVariableBase descriptor && descriptor.Type == dataType)
                .OrderBy(pair => pair.Value.TimeCreated)
                .Select(pair => pair.Value.Name)
                .ToArray();

        /// <summary>
        /// Retrieve the list index of an Id within a specific data type list.
        /// </summary>
        /// <param name="id">The Id to look up.</param>
        /// <param name="dataType">The data type to search up.</param>
        /// <returns>Returns the list Index of the provided Id.</returns>
        public static int ListIndex(Guid id, VariableDataTypes dataType) =>
            Lookup
                .Where(pair => pair.Value is GuildVariableBase descriptor && descriptor.Type == dataType)
                .OrderBy(pair => pair.Value.TimeCreated)
                .ToList()
                .FindIndex(pair => pair.Value.Id == id);

        /// <summary>
        /// Retrieve the Id associated with a list index of a specific data type.
        /// </summary>
        /// <param name="listIndex">The list index to retrieve.</param>
        /// <param name="dataType">The data type to search up.</param>
        /// <returns>Returns the Id of the provided index.</returns>
        public static Guid IdFromList(int listIndex, VariableDataTypes dataType)
        {
            if (listIndex < 0 || listIndex > GetNamesByType(dataType).Length)
            {
                return Guid.Empty;
            }

            return Lookup
                .Where(pair => pair.Value is GuildVariableBase descriptor && descriptor.Type == dataType)
                .OrderBy(pair => pair.Value.TimeCreated)
                .Skip(listIndex)
                .First().Value.Id;
        }
```